### ISSUE TITLE
On larger files all decorations are flickering on (near) every key press

### ISSUE THREAD

**Issue by alexwall** [fabiospampinato/vscode-highlight#105]
I've a file with nearly ~3000 lines and 5 regex definitions. 
Every time gets lager all decorations start flickering on each key press. 

Here is a video:

https://user-images.githubusercontent.com/316469/212502901-f1de3cae-1e3d-4f4a-9209-3fb9e63f408f.mov

It seems to me, that 'highlight' extension is trying to re-decorate the whole file on each small change/key press. 
The interesting part is, that I'm tried the same regular expressions in "[TODO Highlight v2](https://github.com/jgclark/vscode-todo-highlight)" without any flickering. 
I've also tried to comment out all regex except one. The result seemed only a little bit better, still flickering but maybe less frequently. 

The regular expressions are nothing special, here two examples:
```json
"highlight.regexes": {
  "(- \\[x\\])(.*)": {
    "filterLanguageRegex": "markdown",
    "decorations": [
      {
        "filterLanguageRegex": "markdown",
        "color": "#7c7c7c",
      },
      {
        "color": "#7c7c7c"
      }
    ]
  }
  "(.*@today)": {
      "filterLanguageRegex": "markdown",
      "decorations": [
          {
              "backgroundColor": "#006384",
              "fontWeight": "bold",
              "isWholeLine": true
          }
      ]
  },
  "(#wait.*)": {
    "filterLanguageRegex": "markdown",
    "decorations": [
      {
        "color": "#099ade",
        "fontWeight": "bold",
        "isWholeLine": true
      },
      {
        "color": "#099ade"
      }
    ]
  },
}
``` 

I've also tried to find a useful setting at https://code.visualstudio.com/api/references/vscode-api
and tried to experiment with `"highlight.regexFlags"` but without success. 

Is there a way to limit the decoration update to the current line only? Or what does "[TODO Highlight v2](https://github.com/jgclark/vscode-todo-highlight)" could do differently?

**Comment by fabiospampinato** [comment]
> It seems to me, that 'highlight' extension is trying to re-decorate the whole file on each small change/key press.

Unless the edits fall within the bounds of what the extension can optimize that's what happening.

> Is there a way to limit the decoration update to the current line only? 

There's no setting for it, the extension must be able to detect when that's the case.

> Or what does "[TODO Highlight v2](https://github.com/jgclark/vscode-todo-highlight)" could do differently?

No idea, but if the two extensions work similarly either I'm doing something wrong here, or I'm missing some important optimization, or they added new APIs that I'm not aware of that the other extension is using 🤔

**Comment by alexwall** [comment]
The logic of the [relevant lines](https://github.com/fabiospampinato/vscode-highlight/blob/e424ac0b173feafd6d556dd0fcfd553a7f85d3ca/src/decorator.ts#L173) looks similar to [the code of TODO Highlight](https://github.com/jgclark/vscode-todo-highlight/blob/bf54e46147c1cb97da5455e2106eb1903b63f903/src/extension.js#L110). 
By just reviewing the code I was not able to figure out the issue...
My (wild) guess is that the decorate method has an issue to detect single line changes and is always jumping to the [else case here](https://github.com/fabiospampinato/vscode-highlight/blob/e424ac0b173feafd6d556dd0fcfd553a7f85d3ca/src/changes.ts#L35)

**Comment by alexwall** [comment]
Due to debugging I realised, that my guess was wrong...
You evaluate the value for "[areSingleLines](https://github.com/fabiospampinato/vscode-highlight/blob/e424ac0b173feafd6d556dd0fcfd553a7f85d3ca/src/changes.ts#L31)" correctly. 
I'm quite sure that the problem is [here](https://github.com/fabiospampinato/vscode-highlight/blob/e424ac0b173feafd6d556dd0fcfd553a7f85d3ca/src/decorator.ts#L277), 
and to be more specific [here](https://github.com/fabiospampinato/vscode-highlight/blob/e424ac0b173feafd6d556dd0fcfd553a7f85d3ca/src/decorator.ts#L161). 

At [this point ](https://github.com/fabiospampinato/vscode-highlight/blob/e424ac0b173feafd6d556dd0fcfd553a7f85d3ca/src/decorator.ts#L161) you iterate through whole text document and [re-decorate all matches](https://github.com/fabiospampinato/vscode-highlight/blob/e424ac0b173feafd6d556dd0fcfd553a7f85d3ca/src/decorator.ts#L167). So the information of the [affected/changed line](https://github.com/fabiospampinato/vscode-highlight/blob/e424ac0b173feafd6d556dd0fcfd553a7f85d3ca/src/changes.ts#L29) is never used for decoration.

**Comment by fabiospampinato** [comment]
If I remember correctly vscode didn't have APIs for updating decorations 🤔 Like one could only delete them all and add them again. I don't remember exactly how this extension works anymore, if you can find the issue, or the potential new APIs for doing this efficiently, it'd be useful if you could make a PR.

**Comment by alexwall** [comment]
Ok I will try...
I think there is no special new API for this use case. But it should be possible adapt/integrate the logic of the following code snippets based on the information from the `event.contentChanges` object.

1. Detect a change and consume the change event/object:
```typescript
import * as vscode from 'vscode';

export function activate(context: vscode.ExtensionContext) {
    let decoration = vscode.window.createTextEditorDecorationType({
        color: 'red',
        backgroundColor: 'white'
    });

    vscode.workspace.onDidChangeTextDocument(event => {
        let editor = vscode.window.activeTextEditor;
        if (!editor) {
            return;
        }
        let document = editor.document;
        let word = "#tag";
        let ranges = [];
        for (let change of event.contentChanges) {
            let start = change.range.start;
            let end = change.range.end;
            let text = document.getText(new vscode.Range(start, end));
            let match;
            while ((match = new RegExp(word, 'gi').exec(text)) !== null) {
                let start = new vscode.Position(start.line, start.character + match.index);
                let end = new vscode.Position(start.line, start.character + match.index + word.length);
                let wordRange = new vscode.Range(start, end);
                ranges.push(wordRange);
            }
        }
        editor.setDecorations(decoration, ranges);
    });
}
```

2. To expand decoration scope of the upper example to the whole line I would try something like this:
```typescript
let editor = vscode.window.activeTextEditor;
let lineNumber = 5;
let line = editor.document.lineAt(lineNumber);
let lineText = line.text;
console.log(lineText);
```

3. Alternatively expand the range of lines, that should be re-decorated, in case there is a regex that matches multiple lines (respectively: \n):
```typescript
let editor = vscode.window.activeTextEditor;
let lineNumber = 5;
let start = editor.document.lineAt(lineNumber).range.start - 1;
let end = editor.document.lineAt(lineNumber).range.end + 1;
let range = new vscode.Range(start, end);
let lineText = editor.document.getText(range);
console.log(lineText);
```

**Comment by alexwall** [comment]
@fabiospampinato PR is created. I'm looking forward to see it in the next official release asap.

**Comment by alexwall** [comment]
Hi @fabiospampinato, 
thanks that you have accepted my delay-feature-request and also invested time to reimplement it with `_.debounce`. I see that you restored some lines that I've deleted in my original PR. Unfortunately you restored also the root cause for flickering on lager files. Sorry if my PR notes were not particularly meaningful. 

The flickering comes from this line. 
https://github.com/fabiospampinato/vscode-highlight/blob/1ff2a38935066e03e9a15263406a7f6293e37fe6/src/decorator.ts#L209

I guess it was necessary to clear the decoration in previous versions of vscode, but now it's not necessary anymore. 
If you (re-)set new ranges for a specific decoration, previous decoration ranges will be removed automatically. 
This is also why I didn't discovered the flickering in other vscode extensions like  [TODO Highlight](https://github.com/jgclark/vscode-todo-highlight). If you like you can compare your approach with [this lines.](https://github.com/jgclark/vscode-todo-highlight/blob/bf54e46147c1cb97da5455e2106eb1903b63f903/src/extension.js#L147)

Do you mind to remove `Decorator.undecorate();` as a hot fix, or do you prefer a PR?

**Comment by fabiospampinato** [comment]
@alexwall can you post a file and a configuration for this extension that reproduces the problem, so that I can be sure that it doesn't happen or that the situation is at least improved significantly?

I thought about what you are saying, but evidently I seem to be missing something 🤔

**Comment by alexwall** [comment]
@fabiospampinato sure. To reproduce the issue in your extension it's important that you keep changing a `@tag` decoration. 

Here is the TODO Highlight config. The extension has no restrictions like `maxMatches`, which means it re-decorates the whole file (without flickering).  

```json
"todohighlight.keywords": [
    {
        "text": "#mark",
        "color": "black",
        "fontWeight": "bold",
        "overviewRulerColor": "yellow",
        "isWholeLine": true,
        "backgroundColor": "rgba(255,255,0,0.8)",
        "regex": {
            "pattern": "(#mark)\\b"
        },
    },
    {
        "text": "- [x]",
        "color": "grey",
        "isWholeLine": true,
    },
    {
        "text": "@tags", // this is further refined by the regex pattern below
        "color": "#4DE0FF",
        "fontWeight": "bold",
        "regex": {
            "pattern": "(\\s)(@[^@\\s,.;#]+)"
        },
        "isWholeLine": false
    },
  ],
```

And here is an excerpt of my Highlight config. 

```json
"highlight.decorations": {
    "rangeBehavior": 2
},
"highlight.minDelay": 50,
"highlight.regexFlags": "gi",
"highlight.maxMatches": 500,
"highlight.regexes": {
        "(- \\[x\\])(.*)": {
            "filterLanguageRegex": "markdown",
            "decorations": [
                {
                    "filterLanguageRegex": "markdown",
                    "color": "#7c7c7c",
                },
                {
                    "color": "#7c7c7c"
                }
            ]
        },
        "(.*#mark)": {
            "filterLanguageRegex": "markdown",
            "decorations": [
                {
                    "backgroundColor": "#006384",
                    "fontWeight": "bold",
                    "isWholeLine": true
                },
                {
                }
            ]
        },
        "(^|\\s)(@[^@\\s,.;#]+)": {

            "filterLanguageRegex": "markdown",
            "decorations": [
                {
                    // "backgroundColor": "#00ff84",
                },
                {
                    // "overviewRulerColor": "#ff0000",
                    // "backgroundColor": "#006384",
                    // "color": "red",
                    "color": "#4de0ff",
                    "fontWeight": "bold",
                }
            ]
        },
    },
```

The patterns are not 100% equal but isn't an issue in my opinion.

An large example file is also attached. 

[large-file-example.md](https://github.com/fabiospampinato/vscode-highlight/files/10551100/large-file-example.md)

**Comment by fabiospampinato** [comment]
It should be fixed in v1.7.1. Thank you.

**Comment by alexwall** [comment]
Awesome! Thanks!
